File: app/models/computer.py
from app.extensions import db


class Computer(db.Model):
    __tablename__ = "computers"

    idn = db.Column(db.String, primary_key=True) # 2
    data = db.Column(db.JSON, nullable=False)  # 1
     #  installedsoftware   # 3
    installed_software_idns = db.Column(db.JSON, nullable=False) # 4
    network_idn = db.Column(db.JSON, nullable=False)  # 5   
    provides_hardware_quota = db.Column(db.Float, nullable=False) # 6
    software_data_links = db.Column(db.JSON, nullable=False) # 7
    software_idn_mapping = db.Column(db.JSON, nullable=False) # 8
    stored_credentials = db.Column(db.JSON, nullable=False) # 9
    used_hardware_quota = db.Column(db.Float, nullable=False) # 10

    def to_dict(self):
        return {
            "idn": self.idn, # 2
            "data": self.data, # 1
            #  installedsoftware   # 3
            "installed_software_idns": self.installed_software_idns, # 4
            "network_idn": self.network_idn, # 5
            "provides_hardware_quota": self.provides_hardware_quota, # 6
            "software_data_links": self.software_data_links, # 7
            "software_idn_mapping": self.software_idn_mapping, # 8
            "stored_credentials": self.stored_credentials,  # 9
            "used_hardware_quota": self.used_hardware_quota, # 10
        }

File: frontend/src/types/computer.ts
import type { Software } from "./software";

export type Computer = {

  data: string[]; // 1
  idn: string;  // 2   person_group_id:person_index:network_idn

  installed_software: Software[]; // 3
  installed_software_idns: string[]; // 4  person_group_id:person_index:network_idn>cpe_idn#uuid
  network_idn: number[]; // 5
  provides_hardware_quota: number; // 6
  software_data_links: Record<string, string[]>;// 7
  software_idn_mapping: Record<string, string>; // 8
  stored_credentials: Credential[]; // 9
  used_hardware_quota: number;  // 10

  // label?: string;
  device_index?: number;
  suffix?: string;
  previous_idn?: string; 
};
export interface ComputerIdnComponents {
  label?: string;
  device_index?: number;
  suffix?: string;
}
export interface Employee {
  id: number;
  employee_group: string;
  employee_index: number;
  credentials: Credential[];
}

/**
 * Represents a Credential used for accessing software or systems.
 */
export interface Credential {
  idn: string;
  has_root: boolean;
  stored_at_computer_id: string;
  linked_software: Software[];
  linked_employees: Employee[];
}

/**
 * Represents a Data entity, like emails or financial data.
 */
export interface Data {
  idn: string;
  data_definition_idn: string;
  data_type: string;
  database_stored: boolean;
  principal_software_id?: string;
  protection_level: number;
  linked_software: Software[];
}

/**
 * Represents a Software component installed on a computer.
 */
export interface Software_predlozeni {
  idn: string;
  computer_idn: string;
  cpe_idn: string;
  idn_variant: string;
  is_database: string;
  max_client_count: number;
  person_group_id?: string;
  person_index: number;
  requires_hardware_quota: number;
  requires_hardware_quota_per_client: number;
  
  // Relationships
  dependencies: Software[];
  dependency_of: Software[];
  accepts_credentials: Credential[];
  data_links: Data[];
}

export interface FirewallRule {
  idn: string;
  allow: boolean;
  
  // Simplified as arrays of strings for frontend representation
  from_objects: string[];
  to_objects: string[];
}

/**
 * Represents a Network Segment containing multiple computers.
 */
export interface NetworkSegment {
  id: string;
  computers: Computer[];
}

/**
 * Represents the entire system data structure, as stored in the database.
 */
export interface SystemData {
  id: number;
  filename: string;
  data: string; // The raw JSON data as a string
  created_at: string;
}

